### PR TITLE
fixes related to extensions

### PR DESCRIPTION
**Pull request by jangko**
- set RSV bits in frame.encode
- move ExtParams from extutils.nim to types.nim
- remodel extension factory type
- accept/reject extensions offer in server
- offer/accept extensions in client

This PR is cherrypicked from compression experiments, from what we have learned from compression.

### CODE REVIEW

File: ws/types.nim [hunk 1]
```diff
@@ -96,13 +95,21 @@ type
...
+
+  ExtFactoryProc* = proc(
+    isServer: bool,
+    args: seq[ExtParam]): Result[Ext, string] {.
+      gcsafe, raises: [Defect].}
+
+  ExtFactory* = object
+    name*: string
+    factory*: ExtFactoryProc
+    clientOffer*: string
```
Comment (dryajov) [inline ws/types.nim:112]: Just curious, why do we need this?

**Review by dryajov — APPROVED** [review]
LGTM

File: ws/types.nim (hunk 1 above)
Comment (jangko) [reply ws/types.nim:112]: 1. `openArray[(string, ExtFactoryProc)]` does not work for closure proc, we need closure proc.
2. why we need closure proc? we need to capture extension configuration from user
3. keep it ergonomics
4. `clientOffer` is needed before the actual extension creation. the server also can rejects or accepts with configuration different from client offered.

```nim
proc extFactoryMaker(offerPadding: bool, useStdLib: bool): ExtFactory =
   proc factoryProc(.....): Ext =
        # capture extension configuration
        # e.g. `useStdLib`
        
   ExtFactory (
      name : "base64",
      factory: factoryProc,
      clientOffer: "base64; pad=" & $offerPadding  # client configuration
   )
```

Comment (dryajov) [reply ws/types.nim:112]: I was mostly asking about `clientOffer`, but thanks for the broader explanation.

**Pull request merged**